aten/src/ATen/cpu/vec/vec_half.h

```diff
@@ -0,0 +1,59 @@
+#include <ATen/cpu/vec/intrinsics.h>
+
+#ifdef CPU_CAPABILITY_AVX512
+#if defined(__GNUC__)
+#define __at_align__ __attribute__((aligned(64)))
+#elif defined(_WIN32)
+#define __at_align__ __declspec(align(64))
+#else
+#define __at_align__
+#endif
+#else // CPU_CAPABILITY_AVX512
+#if defined(__GNUC__)
+#define __at_align__ __attribute__((aligned(32)))
+#elif defined(_WIN32)
+#define __at_align__ __declspec(align(32))
+#else
+#define __at_align__
+#endif
+#endif // CPU_CAPABILITY_AVX512
+
+namespace at {
+namespace vec {
+// See Note [CPU_CAPABILITY namespace]
+inline namespace CPU_CAPABILITY {
+
+#if defined(CPU_CAPABILITY_AVX2) || defined(CPU_CAPABILITY_AVX512)
+inline uint16_t float_to_half(float val) {
+#if defined(CPU_CAPABILITY_AVX2)
+  __m256 v = _mm256_set1_ps(val);
+  __m128i o =
+      _mm256_cvtps_ph(v, (_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC));
+  return static_cast<std::uint16_t>(_mm_cvtsi128_si32(o));
+#elif defined(CPU_CAPABILITY_AVX512)
+  __m512 v = _mm512_set1_ps(val);
+  __m256i o =
+      _mm512_cvtps_ph(v, (_MM_FROUND_TO_NEAREST_INT | _MM_FROUND_NO_EXC));
+  return static_cast<std::uint16_t>(
+      _mm_cvtsi128_si32(_mm256_castsi256_si128(o)));
+#endif
+}
+
+inline float half_to_float(uint16_t val) {
```

Comment:
I would suggest to make an overload for each pragma if-else for `half2float_scalar` and `float2half_scalar`:
```
#if defined(CPU_CAPABILITY_AVX512)
half2float_scalar { ... }
float2half_scalar { ... }
#elif defined(CPU_CAPABILITY_AVX2)
half2float_scalar { ... }
float2half_scalar { ... }
#else
// calls into the slow path
half2float_scalar { ... }
float2half_scalar { ... }
#endif


```

Reply:
In Half-inl.h : 
```
#if defined(CPU_CAPABILITY_AVX2) || defined(CPU_CAPABILITY_AVX512)
      x(at::vec::float_to_half(value))
#else
      x(detail::fp16_ieee_from_fp32_value(value))
```
If `CPU_CAPABILITY_AVX2` and `CPU_CAPABILITY_AVX512` is not defined, it will use x(detail::fp16_ieee_from_fp32_value(value)).
If we want to define the slow path in vec_half.h we need to include `detail::fp16_ieee_from_fp32_value(value)` in vec_half.h. I think maybe we can leave the slow path in Half-inl.h.